tests/abi/tools/revng-abi-verify/Verify.cpp

```diff
@@ -7,374 +7,228 @@
-toModelRegisters(const State::Registers &Input,
-                 model::Architecture::Values Arch) {
-  State::ModelRegisters Result;
-
-  for (const State::Register &Reg : Input) {
-    auto &Output = Result[model::Register::fromRegisterName(Reg.Name, Arch)];
-    Output.Value = Reg.Value;
-  }
+  revng_assert(!Value.empty());
+  Result.emplace_back(std::move(Value));
 
+  revng_assert(ExpectedChunkCount == Result.size());
   return Result;
 }
 
-static State::Iteration toIteration(const State::SingleRun &SingleRun,
-                                    model::Architecture::Values Architecture) {
-  return State::Iteration{ toModelRegisters(SingleRun.Registers, Architecture),
-                           SingleRun.Stack,
-                           SingleRun.ReturnValues,
-                           SingleRun.Arguments };
-}
+static ExitCode verifyImpl(const abi::FunctionType::Layout &FunctionLayout,
+                           const abi::artifact::FunctionArtifact &Artifact,
+                           model::ABI::Values ABI) {
+  model::Architecture::Values Architecture = model::ABI::getArchitecture(ABI);
+  size_t RegisterSize = model::Architecture::getPointerSize(Architecture);
+  for (size_t Index = 0; Index < Artifact.Iterations.size(); ++Index) {
+    auto &Iteration = Artifact.Iterations[Index];
+
+    // Get the data ready
+    auto ArgRegisters = FunctionLayout.argumentRegisters();
+    auto OrderedRegisterList = abi::orderArguments(std::move(ArgRegisters),
+                                                   ABI);
+    abi::FunctionType::Layout::Argument::StackSpan StackSpan{ 0, 0 };
+    for (const auto &Argument : FunctionLayout.Arguments) {
+      if (Argument.Stack.has_value()) {
+        revng_assert(Argument.Stack->Size != 0);
+        if (StackSpan.Size == 0) {
+          StackSpan = *Argument.Stack;
+        } else {
+          if (Argument.Stack->Offset != StackSpan.Offset + StackSpan.Size)
+            return ExitCode::OnlyContinuousStackArgumentsAreSupported;
+
+          StackSpan.Size += Argument.Stack->Size;
+        }
+      }
+    }
 
-static State::Parsed parse(llvm::StringRef RuntimeArtifact,
-                           model::Architecture::Values Architecture) {
-  llvm::yaml::Input Reader(RuntimeArtifact);
+    // Helper functions.
+    auto FetchRegisterValue = [&](auto Register) -> std::optional<uint64_t> {
+      auto Iterator = Iteration.Registers.find(Register);
+      if (Iterator != Iteration.Registers.end())
+        return Iterator->Value;
+      else
+        return std::nullopt;
+    };
+    auto FetchStackValue = [&](size_t OffsetIndex) -> std::optional<uint64_t> {
+      if (OffsetIndex >= Iteration.Stack.size())
+        return std::nullopt;
+
+      auto &Result = Iteration.Stack[OffsetIndex];
+      revng_assert(Result.Offset >= 0);
+      if (static_cast<uint64_t>(Result.Offset) != OffsetIndex * RegisterSize)
+        return std::nullopt;
+      else
+        return Result.Value;
+    };
```

Comment:
This situation typically calls for a class.

Reply:
It sure does 😅

Reply:
Done: https://github.com/revng/revng/pull/148/commits/709a1641dcd68bc5373d98c9d1f95d42426d98c1.